synapse/http/server.py

```diff
@@ -504,20 +507,38 @@ def __init__(self, request, json_encoder, json_object):
     def start(self):
         self.request.registerProducer(self, False)
 
+    def _send_data(self, data: List[str]):
+        """
+        Send a list of strings as a response to the request.
+        """
+        if not data:
+            return
+        self.request.write("".join(data).encode("utf-8"))
+
     def resumeProducing(self):
         # We've stopped producing in the meantime.
         if not self.request:
             return
 
         # Get the next chunk and write it to the request. Calling write will
         # spin the reactor (and might be re-entrant).
-        try:
-            data = next(self._generator)
-            self.request.write(data.encode("utf-8"))
-        except StopIteration:
-            self.request.unregisterProducer()
-            self.request.finish()
-            self.stopProducing()
+        buffer = []
+        buffered_bytes = 0
+        while buffered_bytes < self.min_chunk_size:
+            try:
+                data = next(self._generator)
+                buffer.append(data)
+                buffered_bytes += len(data)
+            except StopIteration:
+                # Everything is serialized, write any data, then finalize the
+                # producer.
+                self._send_data(buffer)
+                self.request.unregisterProducer()
+                self.request.finish()
+                self.stopProducing()
+                return
+
+        self._send_data(data)
```

Comment:
I think this should be
```suggestion
        self._send_data(buffer)
```

Reply:
Yes, I just realized that and pushed a commit already! Thanks!